packages/material-ui-lab/src/TreeView/descendants.js

```diff
@@ -0,0 +1,217 @@
+    ? React.useLayoutEffect
+    : React.useEffect;
+
+const DescendantContext = React.createContext({});
+
+if (process.env.NODE_ENV !== 'production') {
+  DescendantContext.displayName = 'DescendantContext';
+}
+
+function usePrevious(value) {
+  const ref = React.useRef(null);
+  React.useEffect(() => {
+    ref.current = value;
+  }, [value]);
+  return ref.current;
+}
+
+const noop = () => {};
+
+/**
+ * This hook registers our descendant by passing it into an array. We can then
+ * search that array by to find its index when registering it in the component.
+ * We use this for focus management, keyboard navigation, and typeahead
+ * functionality for some components.
+ *
+ * The hook accepts the element node
+ *
+ * Our main goals with this are:
+ *   1) maximum composability,
+ *   2) minimal API friction
+ *   3) SSR compatibility*
+ *   4) concurrent safe
+ *   5) index always up-to-date with the tree despite changes
+ *   6) works with memoization of any component in the tree (hopefully)
+ *
+ * * As for SSR, the good news is that we don't actually need the index on the
+ * server for most use-cases, as we are only using it to determine the order of
+ * composed descendants for keyboard navigation.
+ */
+export function useDescendant(descendant) {
+  const [, forceUpdate] = React.useState();
+  const {
+    registerDescendant = noop,
+    unregisterDescendant = noop,
+    descendants = [],
+    parentId = null,
+  } = React.useContext(DescendantContext);
+
+  // This will initially return -1 because we haven't registered the descendant
+  // on the first render. After we register, this will then return the correct
+  // index on the following render and we will re-register descendants
+  // so that everything is up-to-date before the user interacts with a
+  // collection.
+  const index = findIndex(descendants, (item) => item.element === descendant.element);
+
+  const previousDescendants = usePrevious(descendants);
+
+  // We also need to re-register descendants any time ANY of the other
+  // descendants have changed. My brain was melting when I wrote this and it
+  // feels a little off, but checking in render and usin  // effect's dependency array works well enough.g the result in the
+  const someDescendantsHaveChanged = descendants.some((d, i) => {
+    return d.element !== previousDescendants?.[i]?.element;
+  });
+
+  // Prevent any flashing
+  useEnhancedEffect(() => {
+    if (!descendant.element) forceUpdate({});
+    registerDescendant({
+      ...descendant,
+      index,
+    });
+    return () => unregisterDescendant(descendant.element);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [
+    registerDescendant,
+    unregisterDescendant,
+    index,
+    someDescendantsHaveChanged,
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    ...Object.values(descendant),
+  ]);
+
+  return { parentId, index };
+}
+
+export function useDescendantsInit() {
```

Comment:
Do we need an abstraction for this?

Reply:
I tried to not alter too much descendants code but I’ll take a look.

Reply:
I tried to not alter too much descendants code but I’ll take a look. We probably don’t need to abstraction - It’s supposed to allow parents to use their descendants but since we register descendants else where we never use it.

Reply:
Removing this is easier in the other PR. So I'm going to make the changes there and keep the style changes in the one.

Reply:
Done - https://github.com/mui-org/material-ui/pull/21592/commits/7ad7db07f8a7a4eed6f7575278ed51d0ec5bd52f